sdk/src/logs/async_batch_log_processor.cc

```diff
@@ -362,8 +145,8 @@ bool AsyncBatchLogProcessor::Shutdown(std::chrono::microseconds timeout) noexcep
   GetWaitAdjustedTime(timeout, start_time);
   // wait for  all async exports to complete and return if timeout reached.
   {
-    std::unique_lock<std::mutex> lock(synchronization_data_->async_export_data_m);
-    synchronization_data_->async_export_waker.wait_for(lock, timeout, [this] {
+    std::unique_lock<std::mutex> lock(export_data_storage_->async_export_data_m);
```

Comment:
I think we should loop wair_for `scheduled_delay_millis_` in a loop until timeout is reached.When `async_export_waker.notify_all()` is called between predicate callback and the real lock operation.This may wait for ever.

Reply:
I am not clear on this. Could you please elaborate on how this might wait forever or share some sample code which fixes this?

Reply:
Thanks, for the clarification. I understood your concern here.
I re-looked into the BatchLogProcessor::ForceFlush at https://github.com/DebajitDas/opentelemetry-cpp/blob/export-changes/sdk/src/logs/batch_log_processor.cc#L112. So it handles this case as follows:
1. when the timeout is set as std::chrono::duration<Rep, Period>::max() or too large, we are waiting on conditiion variable for scheduled_delay_millis_ in loop untill the predicate is true.
2. When the timeout is set as some finite value, we wait till timeout.

Now, if we introduce the above concept in Shutdown, then the predicate in case 1 above would never be true if no callback is called and hence it would block indefinitely. Is this valid then?

Reply:
The problem is the predicate counld be true, but the condition_variable may not be notified any more when shutdown.
Let's looks into the implement of `wait_for` in libstdc++ (GCC 11.2.0).

```cpp
template<typename _Clock, typename _Duration, typename _Predicate>
bool wait_until(unique_lock<mutex>& __lock, const chrono::time_point<_Clock, _Duration>& __atime, _Predicate __p) {
  while (!__p())
    if (wait_until(__lock, __atime) == cv_status::timeout) // The real lock operation
      return __p();
  return true;
}

template<typename _Rep, typename _Period>
cv_status  wait_for(unique_lock<mutex>& __lock, const chrono::duration<_Rep, _Period>& __rtime) {
  using __dur = typename steady_clock::duration;
  return wait_until(__lock, steady_clock::now() + chrono::__detail::ceil<__dur>(__rtime));
}
```

And the threads run as below:

| The thread call `Shutdown()`                             | Background woker thread                                 |                                                              |
| -------------------------------------------------------- | ------------------------------------------------------- | ------------------------------------------------------------ |
| `while (!__p())`                                         | -                                                       | `__p()` returns `false`                                      |
| -                                                        | `export_data->export_ids.push(id);`                     |                                                              |
| -                                                        | `export_data->export_ids_flag[id - 1] = true;`          | `__p()` should return `true` now                             |
| -                                                        | `export_data_storage->async_export_waker.notify_all();` | `notify_all()` is just ignored because it not locked yet.    |
| `if (wait_until(__lock, __atime) == cv_status::timeout)` | -                                                       | `async_export_waker` is locked now, and wait until `__atime` | 

What I mean is just let `async_export_waker` have a change to run `__p()` again and break this loop.